src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc

```diff
@@ -0,0 +1,294 @@
+  EventEngine::ResolvedAddress addr;
+  socklen_t len = EventEngine::ResolvedAddress::MAX_SIZE_BYTES;
+  if (getsockname(socket, const_cast<sockaddr*>(addr.address()), &len) < 0) {
+    return absl::InternalError(
+        absl::StrCat("getsockname:", grpc_core::StrError(errno)));
+  }
+  return EventEngine::ResolvedAddress(addr.address(), len);
+}
+
+}  // namespace
+
+void CFStreamEndpoint::Connect(EventEngine::OnConnectCallback on_connect,
+                               EventEngine::ResolvedAddress addr,
+                               EventEngine::Duration timeout) {
+  peer_address_ = std::move(addr);
+  std::string host_port =
+      grpc_sockaddr_to_string(reinterpret_cast<const grpc_resolved_address*>(
+                                  peer_address_.address()),
+                              true)
+          .value();
+  gpr_log(GPR_INFO, "CFStreamClient::connect, host_port: %s",
+          host_port.c_str());
+  std::string host_string;
+  std::string port_string;
+  grpc_core::SplitHostPort(host_port, &host_string, &port_string);
+  CFStringRef host = CFStringCreateWithCString(NULL, host_string.c_str(),
+                                               kCFStringEncodingUTF8);
+  int port = ResolvedAddressGetPort(peer_address_);
+  CFStreamCreatePairWithSocketToHost(NULL, host, port, &cf_read_stream_,
+                                     &cf_write_stream_);
+
+  CFStreamClientContext cf_context = {0, static_cast<void*>(this), nullptr,
+                                      nullptr, nullptr};
+  CFReadStreamSetClient(
+      cf_read_stream_,
+      kCFStreamEventOpenCompleted | kCFStreamEventHasBytesAvailable |
+          kCFStreamEventErrorOccurred | kCFStreamEventEndEncountered,
+      ReadCallback, &cf_context);
+  CFWriteStreamSetClient(
+      cf_write_stream_,
+      kCFStreamEventOpenCompleted | kCFStreamEventCanAcceptBytes |
+          kCFStreamEventErrorOccurred | kCFStreamEventEndEncountered,
+      WriteCallback, &cf_context);
+  CFReadStreamSetDispatchQueue(cf_read_stream_,
+                               dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0));
+  CFWriteStreamSetDispatchQueue(
+      cf_write_stream_, dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0));
+
+  if (!CFReadStreamOpen(cf_read_stream_)) {
+    auto status = CFErrorToStatus(CFReadStreamCopyError(cf_read_stream_));
+    on_connect(std::move(status));
+    delete this;
+    return;
+  }
+
+  if (!CFWriteStreamOpen(cf_write_stream_)) {
+    auto status = CFErrorToStatus(CFWriteStreamCopyError(cf_write_stream_));
+    on_connect(std::move(status));
+    delete this;
+    return;
+  }
+
+  auto connect_timeout_timer = engine_->RunAfter(timeout, [this]() {
+    open_event_.SetShutdown(absl::DeadlineExceededError("Connect timed out"));
+  });
+
+  open_event_.NotifyOn(new PosixEngineClosure(
+      [this, on_connect = std::move(on_connect),
+       connect_timeout_timer](absl::Status status) mutable {
+        engine_->Cancel(connect_timeout_timer);
```

Comment:
Mmm, it sounds that I'll have to do some tracking or reference counting after all, even if we don't support cancel connection, to prevent from this being deleted because of connection error while the timeout timer is being fired.
I'll need sometime on this.